### DISCUSSION TITLE
Upgrading from 3 to 4 - how to find out if Transaction is active?

### DISCUSSION THREAD

**Discussion by rdehuyss** [micronaut-projects/micronaut-data#2454] [Q&A]
Hi,

I'm Ronald, the creator of [JobRunr](https://github.com/jobrunr/jobrunr), a distributed job scheduling library that supports Micronaut.

I'm currently trying to upgrade to Micronaut 4 but having some difficulties - I tried finding a solution in the docs and the Micronaut Data source code but didn't find it yet.

For some functionality, I need to know whether a transaction is active. The code below worked in v3. What would be the equivalent in v4? 

```java
@Override
    public Connection create(DataSource dataSource) throws SQLException {
        if(TransactionSynchronizationManager.isActualTransactionActive()) {
            try {
                Connection connection = DataSourceUtils.getConnection(dataSource);
                return new MicronautTransactionAwareConnection(connection);
            } catch (CannotGetJdbcConnectionException e) {
                throw (SQLException) e.getCause();
            }
        }
        return dataSource.getConnection();
    }
```

Thanks!

**Comment by dstepanov** [comment]
You might want to inject ‘TransactionOperations’ and call find status or use ‘ConnectionOperations’

**Reply by rdehuyss** [reply]
Thanks @dstepanov for the quick reply! Will this also work in a multithreaded application (is it thread-safe)?

The javadoc does not mention anything about multithreading.

**Reply by dstepanov** [reply]
Yes, the propagated context is stored in the thread local in the same way as before
